src/spam-protection/spam-protection-action-creator.ts

```diff
@@ -0,0 +1,88 @@
+import GoogleRecaptcha from './google-recaptcha';
+import { SpamProtectionAction, SpamProtectionActionType } from './spam-protection-actions';
+import { SpamProtectionOptions } from './spam-protection-options';
+import SpamProtectionRequestSender from './spam-protection-request-sender';
+
+export default class SpamProtectionActionCreator {
+    constructor(
+        private _googleRecaptcha: GoogleRecaptcha,
+        private _requestSender: SpamProtectionRequestSender
+    ) {}
+
+    initialize(options?: SpamProtectionOptions): ThunkAction<SpamProtectionAction, InternalCheckoutSelectors> {
+        return store => Observable.create((observer: Observer<SpamProtectionAction>) => {
+            const spamProtectionElementId = options ? options.containerId : 'spamProtectionContainer';
+
+            if (!options && !document.getElementById(spamProtectionElementId)) {
+                const spamProtectionElement = document.createElement('div');
+                spamProtectionElement.setAttribute('id', spamProtectionElementId);
+                document.body.append(spamProtectionElement);
+            }
+
+            const state = store.getState();
+            const config = state.config.getConfig();
+
+            if (!config) {
+                throw new MissingDataError(MissingDataErrorType.MissingCheckoutConfig);
+            }
+
+            observer.next(createAction(SpamProtectionActionType.InitializeRequested, undefined));
+
+            const recaptchaSitekey = config.storeConfig.checkoutSettings.googleRecaptchaSitekey;
+
+            return this._googleRecaptcha.load(spamProtectionElementId, recaptchaSitekey)
+                .then(() => {
+                    observer.next(createAction(SpamProtectionActionType.InitializeSucceeded));
+                    observer.complete();
+                })
+                .catch(error => {
+                    observer.error(createErrorAction(SpamProtectionActionType.InitializeFailed, error, spamProtectionElementId));
+                });
+        });
+    }
+
+    validate(): ThunkAction<SpamProtectionAction, InternalCheckoutSelectors> {
+        return store => {
+            const state = store.getState();
+            const storeConfig = state.config.getStoreConfig();
+
+            if (!storeConfig) {
+                throw new MissingDataError(MissingDataErrorType.MissingCheckoutConfig);
+            }
+
+            const { isSpamProtectionEnabled } = storeConfig.checkoutSettings;
+
+            if (!isSpamProtectionEnabled) {
+                return empty();
+            }
+
+            return concat(
+                of(createAction(SpamProtectionActionType.ExecuteRequested, undefined)),
+                this._googleRecaptcha.execute()
+                    .pipe(take(1))
+                    .pipe(switchMap(({ error, token }) => {
+                        if (error || !token) {
+                            return throwError(createErrorAction(SpamProtectionActionType.SubmitFailed, error));
```

Comment:
You might want to just throw the error here. And catch it with a `catchError` operator at the end of the observable chain, and re-throw it as an error action. The reason is that there's another potential point of failure (`requestSender.validate`) that you might want to catch and throw the same error action for it.